File: js/samples/04.q.questBot/src/actions/inventoryAction.ts
/* eslint-disable security/detect-object-injection */
import { CardFactory, MessageFactory, TurnContext } from 'botbuilder';
import { Application, ResponseParser } from '@microsoft/teams-ai';
import { ApplicationTurnState, IDataEntities, updateDMResponse } from '../bot';
import { normalizeItemName, searchItemList, textToItemList } from '../items';
import * as responses from '../responses';

/**
 * Adds inventory actions to the given application.
 * @param {Application<ApplicationTurnState>} app The application to add the actions to.
 */
export function inventoryAction(app: Application<ApplicationTurnState>): void {
    app.ai.action('inventory', async (context: TurnContext, state: ApplicationTurnState, data: IDataEntities) => {
        const operation = (data.operation ?? '').toLowerCase();
        switch (operation) {
            case 'update':
                return await updateList(context, state, data);
            case 'list':
                return await printList(app, context, state);
            default:
                await context.sendActivity(`[inventory.${operation}]`);
                return true;
        }
    });
}
/**
 * Updates the user's inventory based on the given data.
 * @param {TurnContext} context The context object for the current turn of conversation.
 * @param {ApplicationTurnState} state The state object for the current turn of conversation.
 * @param {IDataEntities} data The data entities for the current turn of conversation.
 * @returns {Promise<boolean>} A promise that resolves to true if the inventory was updated successfully, or false otherwise.
 */
async function updateList(context: TurnContext, state: ApplicationTurnState, data: IDataEntities): Promise<boolean> {
    const items = Object.assign({}, state.user.value.inventory);
    try {
        // Remove items first
        const changes: string[] = [];
        const remove = textToItemList(data.items);
        for (const item in remove) {
            // Normalize the items name and count
            // - This converts 'coins:1' to 'gold:10'
            const { name, count } = normalizeItemName(item, remove[item]);
            if (!name) {
                continue;
            }

            if (count > 0) {
                // Add the item
                if (Object.prototype.hasOwnProperty.call(items, name)) {
                    items[name] = items[name] + count;
                } else {
                    items[name] = count;
                }
                changes.push(`+${count}(${name})`);
            } else if (count < 0) {
                // remove the item
                const key = searchItemList(name, items);
                if (key) {
                    if (count < items[key]) {
                        changes.push(`-${count}(${key})`);
                        items[key] = items[key] - count;
                    } else {
                        // Hallucinating number of items in inventory
                        changes.push(`-${items[key]}(${key})`);
                        delete items[key];
                    }
                } else {
                    // Hallucinating item as being in inventory
                    changes.push(`-${count}(${name})`);
                }
            }
        }

        // Report inventory changes to user
        const playerName = state.user.value.name.toLowerCase();
        await context.sendActivity(`${playerName}: ${changes.join(', ')}`);

        // Save inventory changes
        state.user.value.inventory = items;
    } catch (err) {
        await updateDMResponse(context, state, responses.dataError());
        return false;
    }

    return true;
}

/**
 * Prints the user's inventory to the conversation.
 * @param {Application<ApplicationTurnState>} app The application object for the current turn of conversation.
 * @param {TurnContext} context The context object for the current turn of conversation.
 * @param {ApplicationTurnState} state The state object for the current turn of conversation.
 * @returns {Promise<boolean>} A promise that resolves to true if the inventory was printed successfully, or false otherwise.
 */
async function printList(
    app: Application<ApplicationTurnState>,
    context: TurnContext,
    state: ApplicationTurnState
): Promise<boolean> {
    const items = state.user.value.inventory;
    if (Object.keys(items).length > 0) {
        state.temp.value.listItems = items;
        state.temp.value.listType = 'inventory';
        const newResponse = await app.ai.completePrompt(context, state, 'listItems');
        if (newResponse) {
            const card = ResponseParser.parseAdaptiveCard(newResponse);
            if (card) {
                await context.sendActivity(MessageFactory.attachment(CardFactory.adaptiveCard(card)));
                state.temp.value.playerAnswered = true;
                return true;
            }
        }

        await updateDMResponse(context, state, responses.dataError());
    } else {
        await updateDMResponse(context, state, responses.emptyInventory());
    }

    return false;
}

File: dotnet/samples/04.q.questBot/Actions/InventoryAction.cs
using AdaptiveCards;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.TeamsAI.AI;
using Microsoft.TeamsAI.AI.Action;
using QuestBot.Models;
using QuestBot.State;

namespace QuestBot.Actions
{
    public partial class QuestBotActions
    {
        [Action("inventory")]
        public async Task<bool> InventoryAsync([ActionTurnContext] ITurnContext turnContext, [ActionTurnState] QuestState turnState, [ActionEntities] Dictionary<string, object> entities)
        {
            ArgumentNullException.ThrowIfNull(turnContext);
            ArgumentNullException.ThrowIfNull(turnState);
            var data = GetData(entities);
            var action = (data.Operation ?? string.Empty).Trim().ToLowerInvariant();
            switch (action)
            {
                case "update":
                    await UpdateListAsync(turnContext, turnState, data);
                    return true;
                case "list":
                    return await PrintListAsync(turnContext, turnState);
                default:
                    await turnContext.SendActivityAsync($"[inventory.{action}]");
                    return true;
            }
        }

        private async Task<bool> PrintListAsync(ITurnContext context, QuestState state)
        {
            var items = state.User!.Inventory;
            if (items != null && items.Count > 0)
            {
                state.Temp!.ListItems = items;
                state.Temp!.ListType = "inventory";
                var newResponse = await _application.AI.CompletePromptAsync(context, state, "listItems", null, default);
                if (!string.IsNullOrEmpty(newResponse))
                {
                    var card = ResponseParser.ParseAdaptiveCard(newResponse);
                    if (card != null)
                    {
                        await context.SendActivityAsync(MessageFactory.Attachment(new Attachment
                        {
                            ContentType = AdaptiveCard.ContentType,
                            Content = card.Card
                        }));
                        state.Temp.PlayerAnswered = true;
                        return true;
                    }
                }

                await UpdateDMResponseAsync(context, state, ResponseGenerator.DataError());
            }
            else
            {
                await UpdateDMResponseAsync(context, state, ResponseGenerator.EmptyInventory());
            }

            return false;
        }

        private static async Task UpdateListAsync(ITurnContext context, QuestState state, DataEntities data)
        {
            var newItems =
                state.User!.Inventory == null ?
                new ItemList() :
                new ItemList(state.User!.Inventory);

            // Remove items first
            var changes = new List<string>();
            var changeItems = ItemList.FromText(data.Items);
            foreach (var removeItem in changeItems)
            {
                // Normalize the items name and count
                // - This converts 'coins:1' to 'gold:10'
                var normalizedItem = ItemList.MapTo(removeItem.Key, removeItem.Value);
                if (string.IsNullOrEmpty(normalizedItem.Key))
                {
                    continue;
                }

                if (normalizedItem.Value > 0)
                {
                    // Add the item
                    if (newItems.ContainsKey(normalizedItem.Key))
                    {
                        newItems[normalizedItem.Key] += normalizedItem.Value;
                    }
                    else
                    {
                        newItems[normalizedItem.Key] = normalizedItem.Value;
                    }
                    changes.Add($"+{normalizedItem.Value} ({normalizedItem.Key})");
                }
                else if (normalizedItem.Value < 0)
                {
                    // remove the item
                    var keyToRemove = newItems.SearchItem(normalizedItem.Key);
                    if (!string.IsNullOrEmpty(keyToRemove))
                    {
                        if (normalizedItem.Value + newItems[keyToRemove] > 0)
                        {
                            changes.Add($"{normalizedItem.Value} ({keyToRemove})");
                            newItems[keyToRemove] += normalizedItem.Value;
                        }
                        else
                        {
                            // Hallucinating number of items in inventory
                            changes.Add($"{newItems[keyToRemove]} ({keyToRemove})");
                            newItems.Remove(keyToRemove);
                        }
                    }
                    else
                    {
                        // Hallucinating item as being in inventory
                        changes.Add($"{normalizedItem.Value} ({normalizedItem.Key})");
                    }
                }
            }

            // Report inventory changes to user
            var playerName = state.User.Name?.ToLowerInvariant() ?? string.Empty;
            await context.SendActivityAsync($"{playerName}: {string.Join(", ", changes)}");

            // Save inventory changes
            state.User.Inventory = newItems;
        }
    }
}
